plugins/woocommerce/includes/wc-cart-functions.php

```diff
@@ -179,21 +182,35 @@ function wc_clear_cart_after_payment() {
 			$order = wc_get_order( $order_id );
 
 			if ( $order instanceof WC_Order && hash_equals( $order->get_order_key(), $order_key ) ) {
-				WC()->cart->empty_cart();
+				$should_clear_cart_after_payment = true;
 			}
 		}
 	}
 
-	if ( WC()->session->order_awaiting_payment > 0 ) {
+	// If the order is awaiting payment, and we haven't already decided to clear the cart, check the order status.
+	if ( WC()->session->order_awaiting_payment > 0 && ! $should_clear_cart_after_payment ) {
 		$order = wc_get_order( WC()->session->order_awaiting_payment );
 
 		if ( $order instanceof WC_Order && $order->get_id() > 0 ) {
-			// If the order has not failed, or is not pending, the order must have gone through.
-			if ( ! $order->has_status( array( 'failed', 'pending', 'cancelled' ) ) ) {
-				WC()->cart->empty_cart();
-			}
+			// If the order status is neither pending, failed, nor cancelled, the order must have gone through.
+			$did_order_succeed = ! $order->has_status( array( 'failed', 'pending', 'cancelled' ) );
+			$should_clear_cart_after_payment = $did_order_succeed;
 		}
 	}
+
+	/**
+	 * Determine whether the cart should be cleared after payment.
+	 *
+	 * @since 9.2.0
+	 * @param bool $should_clear_cart_after_payment Whether the cart should be cleared after payment.
+	 * @param WC_Order $order The order.
+	 * @return bool Whether the cart should be cleared after payment.
```

Comment:
```suggestion
```

No need for a `@return` tag when documenting hooks [[*]](https://developer.wordpress.org/coding-standards/inline-documentation-standards/php/#4-hooks-actions-and-filters).